### DISCUSSION TITLE
Best practices for tree-structured data?

### DISCUSSION THREAD

**Discussion by hcanyz** [rrousselGit/riverpod#3035] [Q&A]
I have a tree structure of data, similar to the following:

```dart
class Node {
  Node({
    required this.id,
    this.parent,
    this.children,
  });

  final String id;
  Node? parent;
  final List<Node>? children;
}
```

Both ```parent``` and ```children``` may be changed

I need to render each Node with minimal rebuilding, are there any best practices for this?

thx~

**Comment by hcanyz** [comment]
I found some similar scenarios like

https://github.com/rrousselGit/riverpod/discussions/1514 Use id instead of nested data

https://github.com/rrousselGit/riverpod/discussions/2876 Use ```select``` to observe only specific properties

But I don't know how to deal with this problem on a tree structure

**Comment by shtse8** [comment]
I am encountering a similar issue with handling complicated, deeply nested objects. I have employed the ID approach, which functions effectively within the notifier, as I have implemented getters to manage the relationships between objects. However, the primary challenge arises within the view, which requires the display of all nested properties and the ability to observe changes. Utilizing `ref.watch(provider)` merely returns the state, but not the notifier. Conversely, `ref.watch(provider.notifier)` does return the notifier but fails to subscribe to any changes. As a result, I find myself needing to re-implement the relationships in the view, given that I cannot access the notifier while simultaneously monitoring changes. Having access to the notifier would allow me to leverage my getters and significantly streamline the process.
